extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -154,45 +154,35 @@ public QueryDefinition makeQueryDefinition(
-
-            // Need to add column names which are present in outputColumnNames and rowSignature but not in bob,
-            // since they need to be present in the row signature for this window stage.
-            for (String columnName : outputColumnNames) {
-              int indexInRowSignature = rowSignature.indexOf(columnName);
-              if (indexInRowSignature != -1 && bob.build().indexOf(columnName) == -1) {
-                ColumnType columnType = rowSignature.getColumnType(indexInRowSignature).get();
-                bob.add(columnName, columnType);
-                log.info("Added column [%s] of type [%s] to row signature for window stage.", columnName, columnType);
-              } else {
-                throw new ISE(
-                    "Found unexpected column [%s] already present in row signature [%s].",
-                    columnName,
-                    rowSignature
-                );
-              }
+      // Iterate over the list of window stages, and add the definition for each window stage to QueryDefinitionBuilder.
+      for (int i = 0; i < stages.size(); i++) {
+        for (WindowOperatorFactory operatorFactory : stages.get(i).getWindowOperatorFactories()) {
+          // Need to add column names which are present in outputColumnNames and rowSignature but not in bob,
+          // since they need to be present in the row signature for this window stage.
+          for (String columnName : operatorFactory.getProcessor().getOutputColumnNames()) {
+            int indexInRowSignature = rowSignature.indexOf(columnName);
+            if (indexInRowSignature != -1 && bob.build().indexOf(columnName) == -1) {
+              ColumnType columnType = rowSignature.getColumnType(indexInRowSignature).get();
+              bob.add(columnName, columnType);
+              log.info("Added column [%s] of type [%s] to row signature for window stage.", columnName, columnType);
+            } else {
+              throw new ISE(
+                  "Found unexpected column [%s] already present in row signature [%s].",
+                  columnName,
+                  rowSignature
+              );
             }
           }
         }
 
         final RowSignature intermediateSignature = bob.build();
         final RowSignature stageRowSignature;
 
-        if (i + 1 == operatorList.size()) {
+        if (i + 1 == stages.size()) {
           stageRowSignature = finalWindowStageRowSignature;
           nextShuffleSpec = finalWindowStageShuffleSpec;
         } else {
-          nextShuffleSpec =
-              findShuffleSpecForNextWindow(operatorList.get(i + 1), queryKitSpec.getNumPartitionsForShuffle());
+          nextShuffleSpec = findShuffleSpecForNextWindow(stages.get(i + 1), queryKitSpec.getNumPartitionsForShuffle());
           if (nextShuffleSpec == null) {
             stageRowSignature = intermediateSignature;
```

Comment:
seems like a possibly problematic way to obtain the rowsignature
it seems to me that `Processor` falls short by returning just `outputColumnNames`

Reply:
I added this in https://github.com/apache/druid/pull/16729/files#diff-1cdba5d733d1ae401b38b93e100543b648fe602e2e434c3e4a4ed23e877a7d66, we had an offline discussion on this at the time of that PR where we agreed and finalized on this approach 😅